### ISSUE TITLE
Problems with use of big timeout (or no timeout at all) in trace service

### ISSUE THREAD

**Issue by AlexPeshkoff** [FirebirdSQL/firebird#7147]
Was originally reported by Pavel Zotov privately.

Bug was found when working with new python testing tool. That tool has no timeout when using trace service.
When working with CS python tool hangs on exit to OS and a number of instances of CS remain running on the server.

It was found that all processes hang in same place - they wait indefinitely for more trace data, and there is no way to stop them. Bug has gone unnoticed when using fbsvcmgr / fbtracemgr cause both have 1 sec timeout. That <1 sec delay when closing utility appears to be not a problem for interactive user and was successfully ignored by all test suites.

It's trivial to reproduce without python slightly modifying out own utility. It was noticed that client also has serious problem - there is no way to interrupt Service::query(). fb_shutdown() also does not help.

**Comment by AlexPeshkoff** [comment]
Applied minimalist fix to FB3.

**Comment by Noremos** [comment]
> Applied minimalist fix to FB3.

Hello Alex.
In classic mode, there seems to be a new issue after your fix. The following messages have begun to appear in my firebird.log:
"Shutting down the server with 1 active connection(s) to 1 database(s), 0 active service(s)"

In the case of the OO API, this happens when not all attachments are detached.
I did a little investigation. Putting sleep(3000ms) before the connection counting function makes the message disappear. looks like a race condition.
I have a simple example to reproduce this.

```
#include <iostream>

#include <stdio.h>
#include <vector>
#include <unistd.h>

#include <ibase.h>
#include <firebird/Interface.h>

using namespace Firebird;
using namespace std;

static IMaster* master = fb_get_master_interface();
static IProvider* prov = master->getDispatcher();

void create_database()
{
	IStatus* st = master->getStatus();
	IUtil* utl = master->getUtilInterface();
	IAttachment* att = NULL;
	IXpbBuilder* dpb = NULL;

	try
	{
		ThrowStatusWrapper status(st);
		dpb = utl->getXpbBuilder(&status, IXpbBuilder::DPB, NULL, 0);
		dpb->insertInt(&status, isc_dpb_page_size, 4 * 1024);

		// create empty database
		att = prov->createDatabase(&status, "/tmp/fbtests.fdb", dpb->getBufferLength(&status),
								   dpb->getBuffer(&status));
		printf("Database /tmp/fbtests.fdb created\n");

		// detach from database
		att->detach(&status);
		att = NULL;
	}
	catch (const FbException& error)
	{
		char buf[256];
		utl->formatStatus(buf, sizeof(buf), error.getStatus());
		fprintf(stderr, "%s\n", buf);
	}

	if (att)
		att->release();
	if (dpb)
		dpb->dispose();
}

IAttachment* attach_database()
{
	IStatus* st = master->getStatus();
	IUtil* utl = master->getUtilInterface();

	IAttachment* att = NULL;
	try
	{
		ThrowStatusWrapper status(st);
		att = prov->attachDatabase(&status, "localhost:/tmp/fbtests.fdb", 0, NULL);
	}
	catch (const FbException& error)
	{
		char buf[256];
		utl->formatStatus(buf, sizeof(buf), error.getStatus());
		fprintf(stderr, "%s\n", buf);
	}

	st->dispose();
	return att;
}

int main()
{
	IStatus* st = master->getStatus();
	ThrowStatusWrapper status(st);

	setenv("ISC_USER", "SYSDBA", 0);
	setenv("ISC_PASSWORD", "masterkey", 0);

	create_database();

	const int a_size = 2;

	std::vector<IAttachment*> arr(a_size);

	IAttachment* arr0 = attach_database();
	IAttachment* arr1 = attach_database();

	arr0->detach(&status);
	arr0 = NULL;
	sleep(1);
	return 0;
}
```

**Comment by Noremos** [comment]
In addition, here stack trace of two threads at the moment the message is printed

```

libEngine12.so!Jrd::JProvider::shutdown(Jrd::JProvider * const this, Firebird::CheckStatusWrapper * status, unsigned int timeout, const int reason)(../src/jrd/jrd.cpp:4425) <- **"Shutting down the serve..."**
libEngine12.so!Firebird::IProviderBaseImpl<Jrd::JProvider, Firebird::CheckStatusWrapper, Firebird::IPluginBaseImpl<Jrd::JProvider, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IReferenceCountedImpl<Jrd::JProvider, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IVersionedImpl<Jrd::JProvider, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IProvider> > > > > > >::cloopshutdownDispatcher(Firebird::IProvider * self, Firebird::IStatus * status, unsigned int timeout, const int reason)(../src/include/firebird/IdlFbInterfaces.h:10645)
libfbclient.so.2!Firebird::IProvider::shutdown<Firebird::CheckStatusWrapper>(Firebird::IProvider * const this, Firebird::CheckStatusWrapper * status, unsigned int timeout, const int reason)(../src/include/firebird/IdlFbInterfaces.h:2098)
libfbclient.so.2!Why::Dispatcher::shutdown(Why::Dispatcher * const this, Firebird::CheckStatusWrapper * userStatus, unsigned int timeout, const int reason)(../src/yvalve/why.cpp:5984)
libfbclient.so.2!fb_shutdown(unsigned int timeout, const int reason)(../src/yvalve/why.cpp:3649)
SRVR_multi_thread(rem_port * main_port, USHORT flags, Thread::Handle integrity_thread)(../src/remote/server/server.cpp:1760)
main(int argc, char ** argv)(../src/remote/server/os/posix/inet_server.cpp:514)

libEngine12.so!Jrd::LockManager::release_shmem(Jrd::LockManager * const this, SLONG owner_offset)(../src/lock/lock.cpp:3191)
libEngine12.so!Jrd::LockManager::LockTableGuard::~LockTableGuard(Jrd::LockManager::LockTableGuard * const this)(../src/lock/lock_proto.h:320)
libEngine12.so!Jrd::LockManager::dequeue(Jrd::LockManager * const this, const SLONG request_offset)(../src/lock/lock.cpp:776)
libEngine12.so!DEQUEUE(Jrd::thread_db * tdbb, Jrd::Lock * lock)(../src/jrd/lck.cpp:169)
libEngine12.so!LCK_release(Jrd::thread_db * tdbb, Jrd::Lock * lock)(../src/jrd/lck.cpp:821)
libEngine12.so!PAGE_LOCK_RELEASE(Jrd::thread_db * tdbb, Jrd::BufferControl * bcb, Jrd::Lock * lock)(../src/jrd/cch.cpp:134)
libEngine12.so!flushAll(Jrd::thread_db * tdbb, USHORT flush_flag)(../src/jrd/cch.cpp:2800)
libEngine12.so!CCH_flush(Jrd::thread_db * tdbb, USHORT flush_flag, TraNumber tra_number)(../src/jrd/cch.cpp:1232)
libEngine12.so!CCH_shutdown(Jrd::thread_db * tdbb)(../src/jrd/cch.cpp:2261)
libEngine12.so!JRD_shutdown_database(Jrd::Database * dbb, const unsigned int flags)(../src/jrd/jrd.cpp:7066)
libEngine12.so!purge_attachment(Jrd::thread_db * tdbb, Jrd::StableAttachmentPart * sAtt, unsigned int flags)(../src/jrd/jrd.cpp:7508)
libEngine12.so!Jrd::JAttachment::freeEngineData(Jrd::JAttachment * const this, Firebird::CheckStatusWrapper * user_status, bool forceFree)(../src/jrd/jrd.cpp:3210)
libEngine12.so!Jrd::JAttachment::detach(Jrd::JAttachment * const this, Firebird::CheckStatusWrapper * user_status)(../src/jrd/jrd.cpp:3170)
libEngine12.so!Firebird::IAttachmentBaseImpl<Jrd::JAttachment, Firebird::CheckStatusWrapper, Firebird::IReferenceCountedImpl<Jrd::JAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IVersionedImpl<Jrd::JAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IAttachment> > > > >::cloopdetachDispatcher(Firebird::IAttachment * self, Firebird::IStatus * status)(../src/include/firebird/IdlFbInterfaces.h:10360)
libfbclient.so.2!Firebird::IAttachment::detach<Firebird::CheckStatusWrapper>(Firebird::IAttachment * const this, Firebird::CheckStatusWrapper * status)(../src/include/firebird/IdlFbInterfaces.h:1982)
libfbclient.so.2!Why::YAttachment::detach(Why::YAttachment * const this, Firebird::CheckStatusWrapper * status)(../src/yvalve/why.cpp:5486)
libfbclient.so.2!Firebird::IAttachmentBaseImpl<Why::YAttachment, Firebird::CheckStatusWrapper, Firebird::IReferenceCountedImpl<Why::YAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IVersionedImpl<Why::YAttachment, Firebird::CheckStatusWrapper, Firebird::Inherit<Firebird::IAttachment> > > > >::cloopdetachDispatcher(Firebird::IAttachment * self, Firebird::IStatus * status)(../src/include/firebird/IdlFbInterfaces.h:10360)
Firebird::IAttachment::detach<Firebird::CheckStatusWrapper>(Firebird::IAttachment * const this, Firebird::CheckStatusWrapper * status)(../src/include/firebird/IdlFbInterfaces.h:1982)
rem_port::disconnect(rem_port * const this, PACKET * sendL, PACKET * receiveL)(../src/remote/server/server.cpp:2965)
process_packet(rem_port * port, PACKET * sendL, PACKET * receive, rem_port ** result)(../src/remote/server/server.cpp:4498)
loopThread()(../src/remote/server/server.cpp:6113)
(anonymous namespace)::ThreadArgs::run((anonymous namespace)::ThreadArgs * const this)(../src/common/ThreadStart.cpp:78)
(anonymous namespace)::threadStart(void * arg)(../src/common/ThreadStart.cpp:94)
libpthread.so.0!start_thread (Unknown Source:0)
libc.so.6!clone (Unknown Source:0)

```

**Comment by aafemt** [comment]
You use embedded mode so your application have to call fb_shutdown() before exit. Besides you didn't close the second attachment so the error in the log is natural.

Nothing of it is related to the topic issue.

**Comment by AlexPeshkoff** [comment]
On 5/31/22 16:08, Dimitry Sibiryakov wrote:
>
> You use embedded mode
>

Is it embedded?

|att = prov->attachDatabase(&status, "localhost:/tmp/fbtests.fdb", 0, NULL);|

I think not.

> Besides you didn't close the second attachment so the error in the log 
> is natural.
>

Here I agree - you really did not close second attachment.
Server certainly closes them instead you - but writes a warning to the log.

> Nothing of it is related to the topic issue.
>

Here also agree.

**Comment by dyemanov** [comment]
Alex, should disconnection due to a broken network link be considered normal or not? Here we have the remote->engine detach happening in parallel with the process shutdown (with reason == fb_shutrsn_no_connection), hence the warning in the log. Wouldn't it be more convenient to have them executed sequentially, because this is what user is likely to expect (server closes the database and then gracefully exists)?

**Comment by AlexPeshkoff** [comment]
Dmitry, the problem here is that querying trace service without timeout 
(may be something else, but that is known one) will never close w/o 
fb_shutdown. I can add check that this is service and call fb_shutdown() 
only in that case. Full fix (for FB4 and hihger) is very hard to 
backport in FB3, therefore suppose some compromise is needed here. 
Hanged CS processes is also not good thing.

**Comment by dyemanov** [comment]
Does it hang inside rdb->rdb_svc->svc_iface->detach() ? If so, we could inform SRVR_multi_thread() before that, so that it calls fb_shutdown() only when needed.

**Comment by AlexPeshkoff** [comment]
Not. We can't invoke svc->detach() cause svc->query() is still running, and Service is not prepared for async detach. Sooner of all it will hang, but I've not tested.

As long as svc->query() hangs in worker thread that thread (obviously) can't complete, and remote server waits for it's completion indefinitely.

**Comment by AlexPeshkoff** [comment]
@stavikpetr Bug already fixed, see https://github.com/FirebirdSQL/firebird/issues/7271

**Comment by pavel-zotov** [comment]
@@@ QA issue @@@
Problem was eliminated in firebird-driver 1.8.0 and firebird-qa plugin 0.17.2.
https://firebird-driver.readthedocs.io/en/latest/ref-core.html#firebird.driver.core.Server.readline_timed
https://firebird-qa.readthedocs.io/en/latest/changelog.html#version-0-17-2
There are lot of tests which use trace, so if some regression will occur then we for sure will encounter this.
There is no much sense to implement test, unless to check for regression in firebird-driver and/or qa-suite.
